### ISSUE TITLE
Test with multiple parameterized TestFixture throws exception upon completion

### ISSUE THREAD

**Issue by jg688** [nunit/nunit#1480]
I'm working on upgrading my tests from 2.6.3 to 3.2.1. I have a (fairly elaborate) suite which has four parameterized TestFixture attributes (the arguments have been sanitized):

```
    [TestFixture("aaaaa", "")]
    [TestFixture("bbbbb", "")]
    [TestFixture("ccccc", "ddddd")]
    [TestFixture("eeeee", "")]
    [Category("ActivationTests")]
    class ConsumerActivationTest : TakesScreenshotOnTestFailure
    {
      ...

```

The call to NUnit is:

`C:\Program Files (x86)\NUnit-3.2.1\bin\nunit3-console.exe  Test.dll --where  cat==ActivationTests --work=C:\App\NUnit3Test\NUnit3.Activation.Consumer
`

The following call also fails in the same manner:

`C:\Program Files (x86)\NUnit-3.2.1\bin\nunit3-console.exe  Test.dll --where class=~FunctionalTests.ActivationTests.Consumer.* --work=C:\App\NUnit3Test\NUnit3.Activation.Consumer
`

All four test suites run to completion, but TestResult.xml only contains the following (I've edited the paths to remove company info):

```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<test-run id="2" testcasecount="1" result="Failed" label="Error" total="1" passed="0" failed="1" inconclusive="0" skipped="0" asserts="0" engine-version="3.2.1.0" clr-version="4.0.30319.34209" start-time="2016-05-05 16:53:28Z" end-time="2016-05-05 17:06:06Z" duration="757.806002">
  <command-line><![CDATA["C:\Program Files (x86)\NUnit-3.2.1\bin\nunit3-console.exe"  App.Test.dll --where class=~FunctionalTests.ActivationTests.Consumer.* --work=C:\App\NUnit3Test\NUnit3.Activation.Consumer]]></command-line>
  <filter>
    <class re="1">FunctionalTests.ActivationTests.Consumer.*</class>
  </filter>
  <test-suite type="Assembly" id="0" name="" fullname="" runstate="NotRunnable" testcasecount="1" result="Failed" label="Error" start-time="2016-05-05 17:06:06Z" end-time="2016-05-05 17:06:06Z" duration="0.001" total="1" passed="0" failed="1" inconclusive="0" skipped="0" asserts="0">
    <failure>
      <message><![CDATA[Exception has been thrown by the target of an invocation.]]></message>
      <stack-trace><![CDATA[
Server stack trace: 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(String assemblyString, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo)
   at System.AppDomain.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.CreateObject(String typeName, Object[] args)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.Run(ITestEventListener listener, String filter)
   at NUnit.Engine.Runners.DirectTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)]]></stack-trace>
    </failure>
  </test-suite>
</test-run>
```

If I comment out three of the four parameterized TestFixtures, everything completes - testResults.xml has the test data output.

Please let me know if there's any more info I can provide. Thanks.

**Comment by vgriph** [comment]
Is there a TestFixture attribute without parameters on the base class TakesScreenshotOnTestFailure and no default constructor in ConsumerActivationTest or do you inherit from ConsumerActivationTest in classes without a constructor matching the signature of the TestFixtrue attributes? I've seen similar errors in that setup. With 2.6 it ignored those TestFixtures it couldn't find a constructor for, but with 3.2.1 it fails to execute any of the tests in the class if one constructor can not be used.

**Comment by jg688** [comment]
No - TakesScreenshotOnTestFailure is a class which consists solely of a TearDown method, the purpose of which is to fire a screen shot when a test fails. Most of my test fixtures inherit this class. ConsumerActivationTest is not inherited by any other class. 

This test runs fine if only one or two TF parameters are used. It only fails with more.

**Comment by ChrisMaddock** [comment]
@jg688 - I'm unable to reproduce this with the below 'minimal example' - this runs fine with NUnit 3.2.1.  Would you be able to work you test down to a minimal example, that you can post in it's entirety? Is there anything unusual in you constructor?

```
using NUnit.Framework;

namespace GeneralIssues
{
    public class ThrowsException
    {
        [TestFixture("a", "b")]
        [TestFixture("b", "c")]
        public class MultipleTestFixtures
        {
            private readonly string _input;

            public MultipleTestFixtures(string input, string wasteage)
            {
                _input = input;
            }

            [Test]
            public void Test()
            {
                Assert.That(_input, Is.EqualTo("a"));
            }
        }
    }
}
```

**Comment by rprouse** [comment]
@jg688 is there anything about your TakeScreenshot code that relies on it being on a specific thread, or type of thread like STA? The threading has changed under the covers a couple of times in NUnit 3.x, so that may be causing you problems if you were inadvertently relying on a threading behaviour.

**Comment by jg688** [comment]
@rprouse Not that I'm aware of - here's the entire class:

```
    public class TakesScreenshotOnTestFailure
    {
        [TearDown]
        public void TakeScreenshotOnTestFailure()
        {
			if (TestContext.CurrentContext.Result.Outcome != ResultState.Success
				&& TestContext.CurrentContext.Result.Outcome != ResultState.Ignored
				&& TestContext.CurrentContext.Result.Outcome != ResultState.Skipped)
                 Driver.TakeScreenshot(TestContext.CurrentContext.Result.Outcome.Status.ToString());
        }
    }

```

...and the method:
```

       public static void TakeScreenshot(string ScreenshotName)
        {
            string timestamp = DateTime.Now.ToString("yyyy-MM-dd-hhmm-ss");
            string fileName = ScreenshotName + TestContext.CurrentContext.Test.Name + ".png";

            if (fileName.Length > 128)
                fileName = fileName.Substring(0, 128);

            string fullPath = Path.Combine(TestContext.CurrentContext.WorkDirectory, Path.GetFileName(fileName));

            try
            {
                Driver.Instance.GetScreenshot().SaveAsFile(fullPath, ImageFormat.Png);
                Logger.WriteError("Saved screenshot as " + fullPath);
            }
            catch (Exception ex)
            {
                Logger.WriteError("Screenshot threw exception: " + ex);
            }
        }

```

**Comment by jg688** [comment]
@ChrisMaddock I've been unable to replicate with your simple example, even with multiple tests and a long namespace. The ConsumerActivationTest constructor is about as simple as it gets:
```

	public ConsumerActivationTest(string PlanSku, string CouponCode)
        {
            _PlanSku = PlanSku;
            _CouponCode = CouponCode;
        }

```
I've been running more tests today and have some additional results to report.

I slightly modified the parameters in my TF to avoid an expected error with the one with two non-null parameters. On running this suite the .xml file (PassedTestResult.xml.txt) was created without problem, however there was a test error as I had mis-typed one of the parameters. I corrected this and re-ran the tests, which then failed to create the .xml (FailedTestResult.xml.txt) with the same exception as originally reported. The four tests **did run to completion** as I see the four accounts they created in our db. The only change between these two runs was the correction of one of the TF params. I've attached the output of both test runs as well as the console out of the failed run. 

One more detail which may be significant: The category "ActivationTests" include two TestFixtures: The one reported above, and a second one which has only one TF parameter. If I run the TF with four parameters by class (--where "class=~Redunda.Gadget.App.Test.Presentation.WebPortal.FunctionalTests.ActivationTests.Consumer.*" or class=~Redunda.Gadget.App.Test.Presentation.WebPortal.FunctionalTests.ActivationTests.Enterrise.*) I do not get the exception. I will try to replicate this scenario in the minimal test example.

.xml from passing test by category:
[PassedTestResult.xml.txt](https://github.com/nunit/nunit/files/257617/PassedTestResult.xml.txt)

.xml from passing test by class:
[TestResultByClass.xml.txt](https://github.com/nunit/nunit/files/257619/TestResultByClass.xml.txt)

.xml from failing test by category:
[FailedTestResult.xml.txt](https://github.com/nunit/nunit/files/257621/FailedTestResult.xml.txt)

**Comment by rprouse** [comment]
Can you run with the `--inprocess` command line flag to take remoting out of the picture? It might give us a more useful exception message and callstack.

**Comment by jg688** [comment]
Yes, I can do this later today.

**Comment by jg688** [comment]
I've gotten inconsistent results, but I can say that if a test run fails without the "--inprocess" flag set, setting the flag and re-running the tests will cause them to pass. I've got a different branch which seems to run the same tests without an issue with the flag *not* set. The differences are:
- The ConsumerActivationTest TF class is now explicitly public
- There's one more test in this TF
- I've removed the NUnitTestAdapter and NUnit.Mocks packages from the build

Why these may have affected this issue, I have no idea. So, unless you want to keep pursuing this (and I'm happy to help in any way I can) I guess this should be marked "could not reproduce".

**Comment by alexSimy** [comment]
I have the same issues. On version 3.2.0 we didn't used to have this problem, but since we've upgraded to version 3.2.1 we get the same error like @jg688 .

```
Exception has been thrown by the target of an invocation.

 Server stack trace: 
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
     at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
     at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
     at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
     at System.Activator.CreateInstance(String assemblyString, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo, StackCrawlMark& stackMark)
     at System.Activator.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo)
     at System.AppDomain.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
     at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
     at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
     at NUnit.Engine.Drivers.NUnit3FrameworkDriver.Run(ITestEventListener listener, String filter)
     at NUnit.Engine.Runners.DirectTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
     at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
     at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)
  
  Exception rethrown at [0]: 
     at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
     at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
     at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
     at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)
  
  Test Run Summary
    Overall result: Failed
    Test Count: 0, Passed: 0, Failed: 0, Inconclusive: 0, Skipped: 0
    Start time: 2016-06-09 09:23:57Z
      End time: 2016-06-09 09:44:41Z
      Duration: 1243.802 seconds
```

If someone can help me in order to fix this issues, please let me know.

**Comment by ChrisMaddock** [comment]
@alexSimy - Do you get an exception if you run with the `--inprocess` flag?

**Comment by alexSimy** [comment]
The thing is that I'm not getting this exception at each run. It is really random. So it may be luck if I'm not getting it.

**Comment by ChrisMaddock** [comment]
@alexSimy This reminds me of #1509, which would make sense for flakey exceptions. See @CharliePoole's post at the end of that thread, and let us know how that goes.

> I believe this is the same error fixed recently by PR #1489. To confirm that, you would have to either wait for a new release or download from our AppVeyor feed at http://ci.appveyor.com/nuget/nunit (take the latest build of master you find there)

**Comment by alexSimy** [comment]
@ChrisMaddock I was testing the `--inprocess` flag several times and it is working. More than 30 attempts and everything passed. Thanks !!!

**Comment by ChrisMaddock** [comment]
@alexSimy Great, but that was meant more for diagnosis, than as a actual solution. :-)

It tells us that the problem is in the actual remoting, and it's not an underlying exception that is being masked. There was a related bug fix that went in recently - try updating to v3.4 when it is released, and see if that solves the problem. (Should be around the end of this week, I believe.)

**Comment by rprouse** [comment]
@ChrisMaddock don't rush me :stuck_out_tongue_winking_eye: 

3.4 is scheduled for the end of the month, maybe end of **next** week if things go well. :smile:

**Comment by ChrisMaddock** [comment]
@rprouse Really?! I thought I'd seen it was this weekend...dammit, wouldn't have been worrying about that memory issue if I knew that! :-p Sorry!!

**Comment by rprouse** [comment]
@ChrisMaddock Really :smile: Sorry if I mis-stated in another thread or wasn't clear.

**Comment by ChrisMaddock** [comment]
@rprouse: Nah, probably me missing reading something again! Cheers for picking up on it! :-)

**Comment by ChrisMaddock** [comment]
@jg688 - Did NUnit 3.4.1 fix this for you?

**Comment by jg688** [comment]
@ChrisMaddock No, sorry. Tried it just now with the -inprocess flag off. Same results:

Server stack trace: 
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(String assemblyString, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo)
   at System.AppDomain.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.CreateObject(String typeName, Object[] args)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.Run(ITestEventListener listener, String filter)
   at NUnit.Engine.Runners.DirectTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
   at System.Runtime.Remoting.Messaging.StackBuilderSink._PrivateProcessMessage(IntPtr md, Object[] args, Object server, Object[]& outArgs)
   at System.Runtime.Remoting.Messaging.StackBuilderSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)]]>

**Comment by rprouse** [comment]
We have made changes in some of these areas. Can you confirm if this is fixed in 3.7.1?

**Comment by jnm2** [comment]
@jg688 We're either going to need you to provide a complete standalone that produces this behavior or we can do nothing but lean on you as a debugging proxy. :D I'm going to close due to lack of response but feel free to reopen if you want to dive in again.
